File: concepts/06_text_binary_and_serialization/03_regular_expressions_and_state/nodejs/test_02_zero_width_and_failure_boundaries.mjs
// 零宽匹配、游标推进与编译失败。
// 共同问题：零宽成功如何避免无限迭代；复用 matcher 是否泄漏游标；
// 非法模式在什么阶段报告。
//
// polyglot-family: text_binary_and_serialization
// polyglot-concept: regular_expressions_and_state
// polyglot-related: languages/nodejs/language/test_014_regular_expressions_flags_state_captures_and_replacement.mjs

import assert from 'node:assert/strict';
import test from 'node:test';

test('matchAll 对零宽结果执行安全推进且不修改原 RegExp', () => {
  const pattern = /(?=.)/g;

  const matches = [...'ab'.matchAll(pattern)];

  assert.deepEqual(matches.map(({ index }) => index), [0, 1]);
  assert.deepEqual(matches.map((match) => match[0]), ['', '']);
  assert.equal(pattern.lastIndex, 0);
});

test('直接 exec 的零宽成功不会自动增加 lastIndex', () => {
  const pattern = /(?=a)/g;

  const first = pattern.exec('a');

  assert.equal(first.index, 0);
  assert.equal(pattern.lastIndex, 0);

  // 手写 while(exec) 必须在空匹配时自行推进，否则会无限循环；matchAll 和字符串算法
  // 使用规范的 AdvanceStringIndex 处理这一边界。
});

test('非法正则源在构造时抛 SyntaxError', () => {
  assert.throws(() => new RegExp('('), SyntaxError);
});

File: concepts/06_text_binary_and_serialization/03_regular_expressions_and_state/julia/test_02_zero_width_and_failure_boundaries.jl
# polyglot-family: text_binary_and_serialization
# polyglot-concept: regular_expressions_and_state
# polyglot-related: languages/julia/language/07_unicode_exceptions_and_resources/
# polyglot-related+: test_053_regular_expressions_and_match_state.jl
#
# 共同问题：零宽匹配如何推进；模式编译失败和无匹配是否使用同一通道。
# 对照观察：零宽结果仍是成功 RegexMatch；无匹配返回 nothing，非法模式在 Regex 构造时抛异常。

using Test

@testset "零宽、无匹配与非法模式是三种结果" begin
    zero_width = match(r"^", "")
    @test zero_width !== nothing
    @test zero_width.match == ""
    @test zero_width.offset == 1
    @test match(r"x", "") === nothing
    @test_throws ErrorException Regex("(")
end
